### PR TITLE
Terrafying loves 2

### PR DESCRIPTION
**Pull request by tombooth**
An attempt to integrate the good things from the terrafying-2 branch without the braking changes. Also sadly not the code reduction yet though

### CODE REVIEW

File: lib/terrafying/generator.rb [hunk 1]
```diff
@@ -9,20 +9,52 @@ module Terrafying
...
+
+      type = @type
+      if @kind != :resource
+        type = [@kind.to_s, @type]
+      end
+
+      key = @key
+      if @kind == :resource && key == nil
+        key = "id"
+      end
```
Comment (DewaldV) [inline lib/terrafying/generator.rb]: We can reuse out `resource?` methods here for some fun:

```ruby
key = @key
resource? && key ||= 'id'
```

Comment (tombooth) [reply lib/terrafying/generator.rb]: I don't feel like drop in clarity is worth the drop of chars

File: lib/terrafying/generator.rb [hunk 2]
```diff
@@ -9,20 +9,52 @@ module Terrafying
...
+      closing_parens = ")" * @fns.count
+      calls = ""
+      if @fns.count > 0
+        calls = @fns.join("(") + "("
+      end
+
+      type = @type
+      if @kind != :resource
+        type = [@kind.to_s, @type]
+      end
```
[the remark this answers is not in the document]
Comment (tombooth) [reply lib/terrafying/generator.rb]: 👍

File: lib/terrafying/generator.rb [hunk 3]
```diff
@@ -9,20 +9,52 @@ module Terrafying
...
+
+      key = @key
+      if @kind == :resource && key == nil
+        key = "id"
+      end
+
+      var = [type, @name, key].flatten.select { |s| s != nil && !s.empty? }.join(".")
```
[the remark this answers is not in the document]
Comment (tombooth) [reply lib/terrafying/generator.rb]: 👍

File: lib/terrafying/generator.rb [hunk 4]
```diff
@@ -9,20 +9,52 @@ module Terrafying
...
+      fn_call("trimspace")
     end
 
     def to_s
-      "${#{@var}}"
+      closing_parens = ")" * @fns.count
+      calls = ""
+      if @fns.count > 0
+        calls = @fns.join("(") + "("
+      end
```
[the remark this answers is not in the document]
Comment (tombooth) [reply lib/terrafying/generator.rb]: 👍

File: lib/terrafying/generator.rb [hunk 5]
```diff
@@ -9,20 +9,52 @@ module Terrafying
...
+          type: "",
+          name:,
+          key: nil,
+          fns: []
+        )
+      @kind = kind
+      @type = type
+      @name = name
+      @key = key
+      @fns = fns
```
[the remark this answers is not in the document]
Comment (tombooth) [reply lib/terrafying/generator.rb]: 👍

File: lib/terrafying/generator.rb [hunk 6]
```diff
@@ -9,20 +9,52 @@ module Terrafying
...
+    end
+
+    def fn_call(fn)
+      Ref.new(kind: @kind, type: @type, name: @name, key: @key, fns: [fn] + @fns)
```
[the remark this answers is not in the document]
Comment (tombooth) [reply lib/terrafying/generator.rb]: 👍

File: lib/terrafying/generator.rb [hunk 7]
```diff
@@ -37,6 +69,20 @@ def ==(other)
       self.to_s == other.to_s
     end
 
+    def [](key)
+      new_key = key
+      if @key != nil
+        new_key = "#{@key}.#{key}"
+      end
```
[the remark this answers is not in the document]
Comment (tombooth) [reply lib/terrafying/generator.rb]: 👍

**Comment by tombooth** [comment]
@samandmoore you might be interested in this. We struggled to switch over to straight terrafying 2 as it required a bunch of reworking. This brings in the best bits of 2, with backwards compatibility. Thanks for your interest!

**Comment by samandmoore** [comment]
Thanks, will take a look!

**Comment by captn3m0** [comment]
Is there documentation for these new features somewhere?

Trying to decide if I should upgrade.

**Pull request merged**